python_interface/design/03_pyHorace_description.md

```diff
@@ -0,0 +1,269 @@
+
+The `pyHorace` implementation, on the other hand will rely on `call_python` with Matlab-side and Python-side wrappers.
+On the Python side, a [`WrappedPythonClass`](https://github.com/mducle/hugo/blob/master/pyHorace/FunctionWrapper.py#L27)
+is used to wrap any Python class such that its objects are stored in a module-global dictionary and only a string reference is passed to Matlab.
+Additional functions, `get_obj_prop` and `call_obj_method` as well as wrapped constructors are defined
+and stored in a module-global dictionary for access by `call_python`.
+On the Matlab side, a [`pyclasswrapper`](https://github.com/mducle/hugo/blob/master/src/pyclasswrapper.m) 
+is used to wrap the these constructors and property access functions so as to present the same interface to the Matlab code as the pure Matlab case.
+The point of these indirections is that only string references rather than objects are passed between Matlab and Python
+but the behaviour is similar to creating objects and accessing their properties and methods directly.
+
+
+# Limitations and future work
+
+There are currently some limitations due to the architecture chosen for `pyHorace` which can be alleviated with more work:
+
+1. The Python syntax for any Matlab function/methods must exactly follow the Matlab syntax.
+2. Matlab functions can only be called as methods of the `Matlab` class and not independently (e.g. `without the m.* prefix`)
+3. Matlab help text cannot be accessed from Python.
+4. MPI routines don't work.
+
+Limitations 1 and 2 come from the choice, outlined in the [first section](#compiled-matlab) to use a gateway function which calls `feval`.
+This can be overcome by writing additional Python wrapper for each Matlab function or class we want to expose to Python.
+Alternatively the `Matlab` and `MatlabProxyObject` classes can be modified to automatically
+convert Python keyword arguments to `('name', value)` pairs to be passed to Matlab if the Matlab code is modified to accept them.
+
+Limitation 3 comes from a restriction of the compiled Matlab output itself:
+Matlab does not allow access to its online documentation system from compiled code.
+The only way to work around this is to save the help text for every Matlab function or class we want to expose to Python to a text file
+and then to either generate wrappers for each (such as that used to overcome limitations 1 and 2)
+or to modify the `Matlab` and `MatlabProxyObject` classes to look up these help text.
+
+Limitation 4 is because the MPI routines needs to launch additional instances of the Matlab interpreter which is not possible from within Python.
+The most straight-forward way to overcome this limitation is to compile Horace as a stand-alone executable, with the `worker_v2.m` function as the gateway function. 
+This executable can then be launched and will communicate with the main process using files or MPI as usual. (This has not been tested).
```

Comment:
This sounds like a critical defect. Is the MPI in Horace or Euphonic/Brille/3rd party? For Horace, does this prevent the current optimization work using MPI?

Reply:
As I mentioned I haven't tested it, and for some reason the current implementation is not working for me on Matlab (on Linux) anyway...

So the first thing we need to do is to create the "compiled" workers. I see that there is some stub code at the moment so this needs to be implemented. I might have time next month for this if Alex is too busy.

Once this is done, I think that in principle it should work with the Python interface as is, or with minimum changes. For the file based stuff we might have to be careful about paths - it seems to use some Matlab settings folder which might not exist or be writable in the compiled Matlab / Python interface. 

For the MPI backend itself... it's rather system dependent unfortunately and the issue is that once we've compiled the Matlab code we can't change it for different systems. I'm not exactly clear on how it's implemented in Herbert but it might need to be refactored to not use `mex` but to load a library (`.so` or `.dll` C-style) which can be swapped out (the Matlab "compiler" includes `mex` files in the compiled archive so they cannot be changed). This way we could accomodate the different MPI implementations without providing a horde of different `mex` files in the distribution... I think I need to talk to Alex about this...